diff --git a/thttpd-2.25b/config.h b/thttpd-2.25b/config.h
index 413e678..7ca84c4 100644
--- a/thttpd-2.25b/config.h
+++ b/thttpd-2.25b/config.h
@@ -342,7 +342,9 @@
 ** individual directories by merely doing a "chmod 711" on them - the
 ** standard Unix file permission to allow file access but disable "ls".
 */
+#ifdef notdef
 #define GENERATE_INDEXES
+#endif
 
 /* CONFIGURE: Whether to log unknown request headers.  Most sites will not
 ** want to log them, which will save them a bit of CPU time.
