diff --git a/src/libopensc/apdu.c b/src/libopensc/apdu.c
index 01a10ca..6f07b75 100644
--- a/src/libopensc/apdu.c
+++ b/src/libopensc/apdu.c
@@ -99,7 +99,11 @@ static int sc_apdu2bytes(sc_context_t *ctx, const sc_apdu_t *apdu,
 			*p = (u8)0x00;
 		break;
 	case SC_APDU_CASE_2_SHORT:
-		*p = (u8)apdu->le;
+		if (apdu->ins == 0xa4 && apdu->p1 == 0 && apdu->p2 == 0) {
+			*p = 0;
+		} else {
+			*p = (u8)apdu->le;
+		}
 		break;
 	case SC_APDU_CASE_2_EXT:
 		if (proto == SC_PROTO_T0)
